treeherder/extract/extract_alerts.py

```diff
@@ -0,0 +1,139 @@
+                            quote_value)
+from jx_mysql.mysql_snowflake_extractor import MySqlSnowflakeExtractor
+from mo_files import File
+from mo_json import (json2value,
+                     value2json)
+from mo_logs import (Log,
+                     constants,
+                     startup)
+from mo_sql import SQL
+from mo_times import (DAY,
+                      Timer)
+from mo_times.dates import (Date,
+                            parse)
+from treeherder.extract import VENDOR_PATH
+
+CONFIG_FILE = (File.new_instance(__file__).parent / "extract_alerts.json").abspath
+
+_keep_import = VENDOR_PATH
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class ExtractAlerts:
+    def run(self, force=False, restart=False, merge=False):
+        # SETUP LOGGING
+        settings = startup.read_settings(filename=CONFIG_FILE)
+        constants.set(settings.constants)
+        Log.start(settings.debug)
+
+        if not settings.extractor.app_name:
+            Log.error("Expecting an extractor.app_name in config file")
+
+        # SETUP DESTINATION
+        destination = bigquery.Dataset(
+            dataset=settings.extractor.app_name, kwargs=settings.destination
+        ).get_or_create_table(settings.destination)
+
+        try:
+            if merge:
+                with Timer("merge shards"):
+                    destination.merge_shards()
+
+            # RECOVER LAST SQL STATE
+            redis = Redis()
+            state = redis.get(settings.extractor.key)
+
+            if restart or not state:
+                state = (0, 0)
+                redis.set(settings.extractor.key, value2json(state).encode("utf8"))
+            else:
+                state = json2value(state.decode("utf8"))
+
+            last_modified, alert_id = state
+            last_modified = parse(last_modified)
+
+            # SCAN SCHEMA, GENERATE EXTRACTION SQL
+            extractor = MySqlSnowflakeExtractor(settings.source)
+            canonical_sql = extractor.get_sql(SQL("SELECT 0"))
+
+            # ENSURE SCHEMA HAS NOT CHANGED SINCE LAST RUN
+            old_sql = redis.get(settings.extractor.sql)
+            if old_sql and old_sql.decode("utf8") != canonical_sql.sql:
+                if force:
+                    Log.warning("Schema has changed")
+                else:
+                    Log.error("Schema has changed")
+            redis.set(settings.extractor.sql, canonical_sql.sql.encode("utf8"))
+
+            # SETUP SOURCE
+            source = MySQL(settings.source.database)
+
+            while True:
+                Log.note(
+                    "Extracting alerts for last_modified={{last_modified|datetime|quote}}, alert.id={{alert_id}}",
+                    last_modified=last_modified,
+                    alert_id=alert_id,
+                )
+                last_year = Date.today() - (
+                    364 * DAY
+                )  # ONLY YOUNG RECORDS CAN GO INTO BIGQUERY
+                get_ids = SQL(
```

Comment:
We should be able to do this with the `ORM` rather than raw `SQL`.  Or does the extractor only work with raw `SQL`?

Reply:
The extractor works with SQL; which is then inserted into a larger piece of sql.  I can try to use the ORM expressions to compose the SQL.

Reply:
@camd   ... this assumes I can phrase the ORM command without sending the SQL; we do not want to pull the list of ids, to just send them back to where they came from.